clippy_lints/src/trivial_default_constructed_types.rs

```diff
@@ -0,0 +1,120 @@
+            && !matches!(qpath, QPath::LangItem(..))
+            && last_path_segment(&qpath).ident.name == kw::Default
+        {
+            let ret_ty = cx
+                .typeck_results()
+                .expr_ty(call)
+                .fn_sig(cx.tcx)
+                .output()
+                .skip_binder()
+                .peel_refs();
+            if let Some(default) = default_value(cx, ret_ty) && !is_from_proc_macro(cx, expr) {
+                span_lint_and_sugg(
+                    cx,
+                    TRIVIAL_DEFAULT_CONSTRUCTED_TYPES,
+                    expr.span,
+                    "constructing a trivial type using `default`",
+                    "try",
+                    default.to_string(),
+                    Applicability::MachineApplicable,
+                );
+            } else if let ty::Tuple(fields) = ret_ty.kind()
+                && let Some(fields_default) = fields.iter()
+                    .map(|field| default_value(cx, field))
+                    .collect::<Option<Vec<&'static str>>>()
+                && !is_from_proc_macro(cx, expr)
+            {
+                let default = if fields.len() == 1 {
+                    // Needs trailing comma to be a single-element tuple
+                    fields_default[0].to_owned() + ","
+                } else {
+                    fields_default.join(", ")
+                };
+
+                span_lint_and_sugg(
+                    cx,
+                    TRIVIAL_DEFAULT_CONSTRUCTED_TYPES,
+                    expr.span,
+                    &format!(
+                        "constructing a {} using `default`",
+                        if fields.is_empty() { "unit" } else { "trivial tuple" },
+                    ),
+                    "try",
+                    format!("({default})"),
+                    Applicability::MachineApplicable,
+                );
+            } else if let ty::Array(ty, len) = ret_ty.kind()
+                && let Some(default) = default_value(cx, *ty)
+                && !is_from_proc_macro(cx, expr)
+            {
+                span_lint_and_sugg(
+                    cx,
+                    TRIVIAL_DEFAULT_CONSTRUCTED_TYPES,
+                    expr.span,
+                    "constructing a trivial array using `default`",
+                    "try",
+                    format!("[{default}; {len}]"),
+                    Applicability::MachineApplicable,
+                );
+            }
+        }
+    }
+}
+
+/// Gets the default value of `ty`.
+fn default_value(cx: &LateContext<'_>, ty: Ty<'_>) -> Option<&'static str> {
+    match ty.kind() {
+        ty::Adt(def, _) => {
+            if is_lang_item_or_ctor(cx, def.did(), LangItem::Option) {
+                return Some("None");
+            }
+
+            None
+        },
+        ty::Bool => Some("false"),
+        ty::Str => Some(r#""""#),
+        ty::Int(_) | ty::Uint(_) => Some("0"),
+        ty::Float(_) => Some("0.0"),
```

Comment:
This might cause an error if the call to default has a qualified type (e.g. `u32::default()`). I would just ignore linting in those cases.

Reply:
I believe in those cases type inference would do its job (like `struct A { x: 0 }` where `x` is `u32`), but I think we can add the suffix to it

Reply:
The suffix will be needed when generics are involved. It's also a problem when type inference can't pick a type, in which case `i32` is assumed.

Reply:
It just occurred to me that the same thing can happen to `Option` as well. You can check if the path to see if it has a type parameter for `Option`.

Reply:
Well, I've made it always add the suffix, does that work? ^^

Reply:
> It just occurred to me that the same thing can happen to `Option` as well. You can check the path to see if it has a type parameter for `Option`.

Oh hmm.. Let me see then

Reply:
Always adding the suffix is fine.